### ISSUE TITLE
Cordova iOS HCP fails if you add a png to your project

### ISSUE THREAD

**Issue by RealHandy** [meteor/meteor#10181]
I've just confirmed that adding a PNG to a project causes iOS hot code push to fail, even though it works fine on Android. See https://forums.meteor.com/t/hot-code-push-worked-on-android-not-on-ios/45279 for this problem related the following iPhone log message, There's an error message for the meteor-webapp plugin, but that seems like it shouldn't be the cause of iOS-only HCP failing:
```
Download failure: Could not link to cached asset: Error Domain=NSCocoaErrorDomain Code=516 ““circle_green.png” couldn’t be linked to “images” because an item with the same name already exists.” UserInfo={NSSourceFilePathErrorKey=/var/mobile/Containers/Data/Application/2DCB2EF6-0F63-4394-944E-6827262A0849/Library/NoCloud/meteor/PartialDownload/app/images/circle_green.png, NSUserStringVariant=(
Link
), NSDestinationFilePath=/var/mobile/Containers/Data/Application/2DCB2EF6-0F63-4394-944E-6827262A0849/Library/NoCloud/meteor/Downloading/app/images/circle_green.png, NSFilePath=/var/mobile/Containers/Data/Application/2DCB2EF6-0F63-4394-944E-6827262A0849/Library/NoCloud/meteor/PartialDownload/app/images/circle_green.png, NSUnderlyingError=0x1c02445f0 {Error Domain=NSPOSIXErrorDomain Code=17 “File exists”}}
ERROR: {"line":36,"column":30,"sourceURL":"http://localhost:12072/plugins/cordova-plugin-meteor-webapp/www/webapp_local_server.js"}
```

**Comment by crapthings** [comment]
do you mean add a png is that you put a png file into public folder ?

**Comment by RealHandy** [comment]
Yes, I added a png to /public/images and redeployed to Galaxy. Android HCP works fine, iOS fails with the error message shown in https://forums.meteor.com/t/hot-code-push-worked-on-android-not-on-ios/45279. Changing my app to use CDN as a workaround isn't something I've tried yet, and that'll probably solve this problem, but I want this bug to go on record, as it has also caused Xcode debugging problems.

**Comment by GeorgeCalvert** [comment]
We're seeing something similar, a la the below.  We're noticing the issue for both font and image files, though which file it is varies with each test.  Furthermore, (1) there is no `CFNetwork....tmp` in our application and (2) the target folder is not part of the codebase we deploy to Cordova.

@abernix @benjamn -- you've been so helpful in the past, perhaps you can get this one prioritized?  Hot codepush is essential for compensating for the lag time that Apple's app review process introduces.

We are actively digging into this, as it's jamming us up with a rollout for a big customer, and will offer up whatever we find.   Thanks :-)

```
2018-08-27 12:37:38.730784 Medssenger[610:92524] Download failure: Could not move downloaded asset: Error Domain=NSCocoaErrorDomain Code=516 "“CFNetworkDownload_b38Uow.tmp” couldn’t be moved to “etl_load_users” because an item with the same name already exists." UserInfo={NSSourceFilePathErrorKey=/private/var/mobile/Containers/Data/Application/745CD66D-BA42-48C2-AE47-EE8CCEE96723/tmp/CFNetworkDownload_b38Uow.tmp, NSUserStringVariant=(
    Move
), NSDestinationFilePath=/var/mobile/Containers/Data/Application/745CD66D-BA42-48C2-AE47-EE8CCEE96723/Library/NoCloud/meteor/Downloading/app/subclassables/etl_load_users/add.png, NSFilePath=/private/var/mobile/Containers/Data/Application/745CD66D-BA42-48C2-AE47-EE8CCEE96723/tmp/CFNetworkDownload_b38Uow.tmp, NSUnderlyingError=0x1740593b0 {Error Domain=NSPOSIXErrorDomain Code=17 "File exists"}}
```

**Comment by RealHandy** [comment]
Okay, I've debugged the problem and opened a specific issue here: https://github.com/meteor/cordova-plugin-meteor-webapp/issues/56 (as indicated in the reference above)

**Comment by crapthings** [comment]
we have same issue as @RealHandy ,  we just move all public files into cdn, and HCP works fine even from meteor 1.4.2.x to 1.7.0.1

and don't forget to put COMPAT_VERSION before you build.

export METEOR_CORDOVA_COMPAT_VERSION_ANDROID='your old hash here'
export METEOR_CORDOVA_COMPAT_VERSION_IOS='your old hash here'

**Comment by hexsprite** [comment]
@crapthings so the COMPAT env vars are a must or HCP will break?

**Comment by RealHandy** [comment]
@hexsprite I haven't had any issues with HCP not working if the hashes don't change. I don't yet provide them to the build b/c they're staying the same.

**Comment by GeorgeCalvert** [comment]
Our diagnostics pinned down the error -- at least as we experience it -- to be caused when we rename a directory containing an image.  In our case, these were ancestor directories in the source tree (i.e. outside public).  And from @RealHandy's discovery -- thanks! -- adding a directory is also a problem.  (It now looks as if any issue with font-files was a false positive.)

We renamed some directories in upgrading from 1.6 to 1.7.   We telegraph upgrades to mobile by releasing the Cordova change early, allowing evergreen devices to update before we push all changes to the server (which then distributes the new functionality via HCP).  This gave us a 1.7 client and a 1.7-based server with changes that included the renamed directories.  That broke HCP -- but distributing a new 1.7 client, one with the new directory names, overcame the issue.  

I'll see if we can't build on @RealHandy's work in [cordova-plugin-meteor-webapp#56](https://github.com/meteor/cordova-plugin-meteor-webapp/issues/56).  It's a big headache if we're gonna see this bug for every new directory containing a PNG, even when those directories are outside the source we send to mobile devices.

**Comment by RealHandy** [comment]
https://github.com/meteor/cordova-plugin-meteor-webapp/issues/56#issuecomment-417693985

**Comment by heberallred** [comment]
Does anyone know if this issue will be resolved in Meteor 1.8.1?

**Comment by RealHandy** [comment]
@GeorgeCalvert The additional update made to the fork yesterday (see https://github.com/meteor/cordova-plugin-meteor-webapp/pull/59) may fix the problem you mentioned here.

**Comment by RealHandy** [comment]
Current activity for this issue is at #10277

**Comment by RealHandy** [comment]
Closing this, as it's fixed, but also #10277 ended up being where all the action was.
